src/browser.ts

```diff
@@ -0,0 +1,214 @@
+import {GaxCall, GRPCCall} from './apitypes';
+import {Descriptor} from './descriptor';
+import {createApiCall as _createApiCall} from './createApiCall';
+
+export {PathTemplate} from './pathTemplate';
+export {CallSettings, constructSettings, RetryOptions} from './gax';
+
+export {
+  BundleDescriptor,
+  LongrunningDescriptor,
+  PageDescriptor,
+  StreamDescriptor,
+} from './descriptor';
+
+export {StreamType} from './streamingCalls/streaming';
+
+export const ALL_SCOPES: string[] = [];
+lro.ALL_SCOPES = ALL_SCOPES;
+
+interface CancelHandler {
+  canceller: AbortController;
+  cancelRequested: boolean;
+}
+
+export class GrpcClient {
+  auth: GoogleAuth;
+  promise?: PromiseConstructor;
+
+  constructor(options: GrpcClientOptions = {}) {
+    if (!options.auth) {
+      throw new Error(
+        'You need to pass auth instance to gRPC-fallback client. Use OAuth2Client from google-auth-library.'
+      );
+    }
+    this.auth = options.auth;
+    this.promise = 'promise' in options ? options.promise! : Promise;
+  }
+
+  loadProto(jsonObject) {
+    const rootObject = protobuf.Root.fromJSON(jsonObject);
+    return rootObject;
+  }
+
+  private getServiceMethods(service: protobuf.Service) {
+    const methods = Object.keys(service.methods);
+
+    // Creating an array of all methods within the stub with correct case (i.e, first letter is lower-case)
+    const methodsCorrectCase = methods.map(method => {
+      return method[0].toLowerCase() + method.substring(1);
+    });
+
+    return methodsCorrectCase;
+  }
+
+  constructSettings(
+    serviceName: string,
+    clientConfig: gax.ClientConfig,
+    configOverrides: gax.ClientConfig,
+    headers: OutgoingHttpHeaders
+  ) {
+    return gax.constructSettings(
+      serviceName,
+      clientConfig,
+      configOverrides,
+      Status,
+      this.promise
+    );
+  }
+
+  async createStub(service: protobuf.Service, opts: ClientStubOptions) {
+    const authHeader = await this.auth.getRequestHeaders();
+    function serviceClientImpl(method, requestData, callback) {
+      let cancelController, cancelSignal;
+      if (typeof AbortController !== 'undefined') {
+        cancelController = new AbortController();
+        cancelSignal = cancelController.signal;
+      }
+      const cancelHandler: CancelHandler = {
+        canceller: cancelController,
+        cancelRequested: false,
+      };
+      const headers = Object.assign({}, authHeader);
+      headers['Content-Type'] = 'application/x-protobuf';
+      headers['User-Agent'] = 'testapp/1.0';
+
+      const grpcFallbackProtocol = opts.protocol || 'https';
+      let servicePath = opts.servicePath;
+      if (!servicePath) {
+        if (service.options && service.options['(google.api.default_host)']) {
+          servicePath = service.options['(google.api.default_host)'];
+        } else {
+          callback(new Error('Service path is undefined'));
+          return;
+        }
+      }
+      const servicePort = opts.port || 443;
```

Comment:
FYI: There is a chance that the `google.api.default_host` value has a port in it, in which case it must be honored by clients. [Example in showcase](https://github.com/googleapis/gapic-showcase/blob/74e71466e0d5badb3167900a553a6afea62b6e3d/schema/google/showcase/v1beta1/echo.proto#L38) and this is done in [gapic-generator](https://github.com/googleapis/gapic-generator/blob/13148087d12bd6612f470900f9f20b254bf162e3/src/main/java/com/google/api/codegen/config/ProductServiceConfig.java#L34-L46), gapic-generator-python, gapic-generator-go.

We might want to revisit this logic in conjunction with the use of the option above.

Reply:
Agreed, let's parse the hostname for the port (regular expression again :) )